btree/cache.go

```diff
@@ -0,0 +1,111 @@
+package btree
+
+type item[K any, P comparable, V any] struct {
+	dirty bool
+	node  *Node[K, P, V]
+}
+
+type lru[P comparable] struct {
+	ptr  P
+	next *lru[P]
+}
+
+type cache[K any, P comparable, V any] struct {
+	size   int
+	target int
+	store  Storer[K, P, V]
+	items  map[P]*item[K, P, V]
+	head   *lru[P]
+	tail   *lru[P]
+
+	hits int
+	miss int
+}
+
+func cachefor[K any, P comparable, V any](store Storer[K, P, V], order int) *cache[K, P, V] {
+	target := order
+	return &cache[K, P, V]{
+		target: target,
+		store:  store,
+		items:  make(map[P]*item[K, P, V], target),
+	}
+}
+
+func (c *cache[K, P, V]) cache(ptr P, node *Node[K, P, V]) {
+	c.size++
+	c.items[ptr] = &item[K, P, V]{node: node}
+	lru := &lru[P]{ptr: ptr}
+
+	if c.head == nil {
+		c.head = lru
+		c.tail = lru
+	} else {
+		c.tail.next = lru
```

Comment:
this seems weird:

c.tail.next = lru
c.tail = lru

wasn't c.tail.next expected to be a nil pointer ?

Reply:
ok, it's me, I forgot how linked lists worked :-)